File: 1110.delete-nodes-and-return-forest.py
#
# @lc app=leetcode id=1110 lang=python3
#
# [1110] Delete Nodes And Return Forest
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        res = []
        to_delete = set(to_delete)
        q = deque([(root, True, root.val in to_delete)])

        while q:
            node, is_root, marked = q.popleft()
            if is_root and (not marked):
                res.append(node)
            if node.left:
                is_deleted = node.left.val in to_delete
                q.append((node.left, marked, is_deleted))
                if is_deleted:
                    node.left = None
            if node.right:
                is_deleted= node.right.val in to_delete
                q.append((node.right, marked, is_deleted))
                if is_deleted:
                    node.right = None

        return res

        # def dfs(node, is_root, marked):
        #     if is_root and (not marked):
        #         res.append(node)
        #     if node.left:
        #         is_deleted = node.left.val in to_delete
        #         dfs(node.left, marked, is_deleted)
        #         if is_deleted:
        #             node.left = None
        #     if node.right:
        #         is_deleted= node.right.val in to_delete
        #         dfs(node.right, marked, is_deleted)
        #         if is_deleted:
        #             node.right = None

        # dfs(root, True, root.val in to_delete)


# @lc code=end

File: 1110.delete-nodes-and-return-forest.rs
/*
 * @lc app=leetcode id=1110 lang=rust
 *
 * [1110] Delete Nodes And Return Forest
 */

// @lc code=start
// Definition for a binary tree node.
use std::{
    cell::RefCell,
    collections::{HashSet, VecDeque},
    rc::Rc,
};

// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }

// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

impl Solution {
    pub fn del_nodes(
        root: Option<Rc<RefCell<TreeNode>>>,
        to_delete: Vec<i32>,
    ) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        let mut res = vec![root.clone()];
        let to_delete = to_delete.into_iter().collect::<HashSet<i32>>();
        let mut q = VecDeque::new();

        let root_val = root.as_ref().unwrap().borrow().val;
        q.push_back((root, true, to_delete.contains(&root_val)));

        while let Some((node, is_root, marked)) = q.pop_front() {
            if is_root && !marked {
                res.push(node.clone());
            }
            let mut parent = node.as_ref().unwrap().borrow_mut();

            if let Some(left) = parent.left.take() {
                let is_deleted = to_delete.contains(&left.borrow().val);
                q.push_back((Some(left), marked, is_deleted));
                if is_deleted {
                    parent.left = None;
                }
            }
            if let Some(right) = parent.right.take() {
                let is_deleted = to_delete.contains(&right.borrow().val);
                q.push_back((Some(right), marked, is_deleted));
                if is_deleted {
                    parent.right = None;
                }
            }
        }
        res
    }
}
// @lc code=end
